File: app/src/main/java/org/eclipsesoundscapes/service/NotificationReceiver.java
package org.eclipsesoundscapes.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import org.eclipsesoundscapes.R;
import org.eclipsesoundscapes.data.DataManager;
import org.eclipsesoundscapes.data.SharedPrefsHelper;
import org.eclipsesoundscapes.model.MediaItem;
import org.eclipsesoundscapes.ui.main.MainActivity;
import org.eclipsesoundscapes.ui.media.MediaPlayerActivity;

import static org.eclipsesoundscapes.service.NotificationScheduler.ECLIPSE_FIRST_CONTACT;
import static org.eclipsesoundscapes.service.NotificationScheduler.ECLIPSE_LAUNCH_MEDIA;
import static org.eclipsesoundscapes.service.NotificationScheduler.ECLIPSE_NOTIFICATION_TYPE;
import static org.eclipsesoundscapes.service.NotificationScheduler.showNotification;

/*
 * Displays any scheduled notification or re-schedules them after system reboot
 */
public class NotificationReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        if (context == null) return;

        final boolean isLiveEnabled = context.getResources().getBoolean(R.bool.live_experience_enabled);
        if (!isLiveEnabled) {
            return;
        }

        if (intent.getAction() != null && intent.getAction().equalsIgnoreCase(Intent.ACTION_BOOT_COMPLETED)) {

            SharedPrefsHelper sharedPrefsHelper = new SharedPrefsHelper(context);
            DataManager dataManager = new DataManager(sharedPrefsHelper);

            // Reset reminders.
            NotificationScheduler.scheduleNotifications(context,
                    dataManager.firstContactDate(),
                    dataManager.totalityDate());
            return;
        }

        // Trigger notification
        if (intent.hasExtra(ECLIPSE_NOTIFICATION_TYPE)) {
            final String type = intent.getStringExtra(ECLIPSE_NOTIFICATION_TYPE);
            final boolean launchMedia = intent.getBooleanExtra(ECLIPSE_LAUNCH_MEDIA, false);
            final boolean isFirstContact = type.equals(ECLIPSE_FIRST_CONTACT);

            // reminder
            if (!launchMedia){
                final String description = isFirstContact ?
                        context.getString(R.string.first_contact_reminder) : context.getString(R.string.totality_reminder);
                Intent mediaIntent = new Intent(context, MainActivity.class);
                showNotification(context, MainActivity.class, type, false,
                        context.getString(R.string.app_name), description, mediaIntent);
                return;
            }

            // launch media player on notification press
            Intent mediaIntent = new Intent(context, MediaPlayerActivity.class);
            mediaIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK |
                    Intent.FLAG_ACTIVITY_SINGLE_TOP);

            if (isFirstContact){
                mediaIntent.putExtra(MediaPlayerActivity.EXTRA_MEDIA, new MediaItem(R.drawable.eclipse_first_contact, R.string.first_contact, R.string.audio_first_contact_full, R.raw.first_contact_full));
                mediaIntent.putExtra(MediaPlayerActivity.EXTRA_LIVE, true);
                showNotification(context, MediaPlayerActivity.class, type, true,
                        context.getString(R.string.first_contact_begun), context.getString(R.string.tap_listen), mediaIntent);
            } else {
                mediaIntent.putExtra(MediaPlayerActivity.EXTRA_MEDIA, new MediaItem(R.drawable.eclipse_bailys_beads, R.string.bailys_beads, R.string.bailys_beads_short, R.raw.realtime_eclipse_shorts_saas));
                mediaIntent.putExtra(MediaPlayerActivity.EXTRA_LIVE, true);
                showNotification(context, MediaPlayerActivity.class, type, true,
                        context.getString(R.string.totality_begun), context.getString(R.string.tap_listen), mediaIntent);
            }
        }
    }
}

File: app/src/main/java/org/eclipsesoundscapes/service/NotificationReceiver.kt
package org.eclipsesoundscapes.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.work.ExistingWorkPolicy
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import dagger.hilt.android.AndroidEntryPoint
import org.eclipsesoundscapes.model.Eclipse
import org.eclipsesoundscapes.model.EclipseType
import org.eclipsesoundscapes.service.NotificationScheduler.NOTIFICATION_ECLIPSE_EVENT
import org.eclipsesoundscapes.service.NotificationScheduler.NOTIFICATION_ECLIPSE_TYPE
import org.eclipsesoundscapes.service.NotificationScheduler.NOTIFICATION_LAUNCH_MEDIA
import org.eclipsesoundscapes.workers.RescheduleNotificationsWorker

/*
 * Displays any scheduled notification or re-schedules them after system reboot
 */
@AndroidEntryPoint
class NotificationReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != null && intent.action.equals(Intent.ACTION_BOOT_COMPLETED, ignoreCase = true)) {
            rescheduleNotifications(context)
        } else {
            showNotification(context, intent)
        }
    }

    private fun showNotification(context: Context, intent: Intent) {
        val eclipseName = intent.getStringExtra(NOTIFICATION_ECLIPSE_EVENT)
        val typeName = intent.getStringExtra(NOTIFICATION_ECLIPSE_TYPE)
        if (eclipseName.isNullOrEmpty() || typeName.isNullOrEmpty()) {
            return
        }

        val eclipse = enumValueOf<Eclipse>(eclipseName)
        val type = enumValueOf<EclipseType>(typeName)
        val showMedia = intent.getBooleanExtra(NOTIFICATION_LAUNCH_MEDIA, false)

        NotificationScheduler.showNotification(context, eclipse, type, showMedia)
    }

    private fun rescheduleNotifications(context: Context) {
        val workRequest = OneTimeWorkRequestBuilder<RescheduleNotificationsWorker>()
            .build()

        WorkManager.getInstance(context).enqueueUniqueWork(
            RescheduleNotificationsWorker.NOTIFICATIONS_WORK_NAME,
            ExistingWorkPolicy.REPLACE,
            workRequest)
    }
}